module/zcommon/zfs_prop.c

```diff
@@ -402,7 +402,7 @@ zfs_prop_init(void)
 	    "skein[,verify] | edonr,verify",
 	    "DEDUP", dedup_table, sfeatures);
 	zprop_register_index(ZFS_PROP_COMPRESSION, "compression",
-	    ZIO_COMPRESS_DEFAULT, PROP_INHERIT,
+	    ZIO_COMPRESS_ON, PROP_INHERIT,
```

Comment:
> (If you're wondering why I didn't just change ZIO_COMPRESS_DEFAULT, I tried that first; my recollection is that I got some spectacular crashes.)

Let's retry with your original change to set `ZIO_COMPRESS_DEFAULT` to `ZIO_COMPRESS_ON`.  Near as I can tell, `ZIO_COMPRESS_DEFAULT` is used in only this one place, it's hard to imagine how it could cause what you hit. 

```diff
-#define        ZIO_COMPRESS_DEFAULT            ZIO_COMPRESS_OFF
+#define        ZIO_COMPRESS_DEFAULT            ZIO_COMPRESS_ON
```